wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::OutputSectionNode,
+        ])
+    }
+}
+
+/// Check declaration name for type prefixes or suffixes
+fn check_decl_name(
+    state: &mut Diagnostics,
+    decl: &Decl,
+    exceptable_nodes: &Option<&'static [SyntaxKind]>,
+) {
+    let name = decl.name();
+    let name_str = name.as_str();
+    let name_lower = name_str.to_lowercase();
+    
+    // Get the declaration type
+    let ty = decl.ty();
+    
+    // Extract base type name(s) using AST methods
+    let mut type_names = Vec::new();
+    
+    // Add the full type name
+    type_names.push(ty.to_string());
+    
+    // If it's an array, add the element type
+    if let Some(array_type) = ty.as_array_type() {
+        // Add the element type 
+        type_names.push(array_type.element_type().to_string());
+    }
+    
+    // If it's a map, add the key and value types
+    if let Some(map_type) = ty.as_map_type() {
+        let (key, value) = map_type.types();
+        type_names.push(key.to_string());
+        type_names.push(value.to_string());
+    }
+    
+    // If it's a pair, add the left and right types
+    if let Some(pair_type) = ty.as_pair_type() {
+        let (left, right) = pair_type.types();
+        type_names.push(left.to_string());
+        type_names.push(right.to_string());
+    }
+    
+    // If it's a primitive type, use that directly
+    if let Some(primitive_type) = ty.as_primitive_type() {
+        // Convert the PrimitiveTypeKind to a string
+        let kind_str = match primitive_type.kind() {
+            PrimitiveTypeKind::Boolean => "Boolean",
+            PrimitiveTypeKind::Integer => "Int",
+            PrimitiveTypeKind::Float => "Float",
+            PrimitiveTypeKind::String => "String",
+            PrimitiveTypeKind::File => "File", 
+            PrimitiveTypeKind::Directory => "Directory",
+        };
+        type_names.push(kind_str.to_string());
+    }
+    
+    // If it's a type reference, add the reference name
+    if let Some(type_ref) = ty.as_type_ref() {
+        type_names.push(type_ref.name().as_str().to_string());
+    }
+    
+    // Remove optional markers (?) from type names
+    let type_names: Vec<String> = type_names
+        .into_iter()
+        .map(|t| t.trim_end_matches('?').to_string())
+        .collect();
+    
+    // Check each type name against the declaration name
+    for type_name in &type_names {
+        let type_lower = type_name.to_lowercase();
+        
+        // Skip if the type name is too short (to avoid false positives)
+        if type_lower.len() < 2 {
+            continue;
+        }
+        
+        // Check if the declaration name contains the type name
+        if name_lower.starts_with(&type_lower) || 
+           name_lower.ends_with(&type_lower) || 
+           name_lower.contains(&type_lower) {
```

Comment:
@a-frantz - I'm not sure about the `contains` check. `int` in particularly is likely to be a large problem. We could maybe do a `contains` check on the other types though.

Reply:
That said, the `starts_with` check is also problematic for `int`. Anything with the `inter-` or `intra-` prefixes is going to get flagged.

Reply:
One example from `workflows`: https://github.com/stjudecloud/workflows/blob/1539237af8d5fdaf34f1709d025e7e518caebca1/workflows/reference/gatk-reference.wdl#L65